wandb/sdk/wandb_run.py

```diff
@@ -1959,16 +1997,160 @@ def _on_ready(self) -> None:
         # object is about to be returned to the user, don't let them modify it
         self._freeze()
 
+    def _log_job(self) -> None:
+        artifact = None
+        input_types = TypeRegistry.type_of(self.config.as_dict()).to_json()
+        output_types = TypeRegistry.type_of(self.summary._as_dict()).to_json()
+
+        import pkg_resources
+
+        installed_packages_list = sorted(
+            f"{d.key}=={d.version}" for d in iter(pkg_resources.working_set)
+        )
+
+        for job_creation_function in [
+            self._create_repo_job,
+            self._create_artifact_job,
+            self._create_image_job,
+        ]:
+            artifact = job_creation_function(
+                input_types, output_types, installed_packages_list
+            )
+            if artifact:
+                artifact.wait()
+                if not artifact.metadata:
+                    artifact.metadata["config_defaults"] = self.config.as_dict()
+                    artifact.save()
+                break
+            else:
+                logger.info(
+                    f"Failed to create job using {job_creation_function.__name__}"
+                )
+
+    def _create_repo_job(
```

Comment:
nit: for `_create_repo_job` and `_create_artifact_job` would be nice to abstract out the artifact creation if possible, similar to `job_creation_function` then have these call out to those functions.

Reply:
Im a little confused about this comment actually. Do you mean have a generic function to create the artifact and add the files? Why shouldn't _create_image_job do this?